### PR TITLE
Retry obs_reset_video with default params if user ones fail

### PR DESCRIPTION
**Pull request by eugen15**
### Description
- Moved `obs_reset_video` call to a separate wrapper method
- Extended the `OBS_service::resetVideoContext` method to check for errors and retry initialization with default video configuration if the user one fails
- Added the `OBS_service::keepFallbackVideoConfig` method to copy the default video configuration to the user one, then save it to `basic.ini`; the method is used by `OBS_service::resetVideoContext` if retrying is successful

### Motivation and Context
Theoretically, this change may help when a user has a corrupted config or wrong settings.

### How Has This Been Tested?
1. Change one of the `basic.ini` settings to -1 or 65535
```
[Video]
BaseCX=1920
BaseCY=1080
OutputCX=1920
OutputCY=1080
```
2. Change one of non-**Video** settings to something non-default but standard (i.e. a correct value)
3. Start Streamlabs Desktop (without the changes it will not start correctly)
4. Check logs in `slobs-client\node-obs\logs` for the following messages:
- `Prepared obs_video_info:` for the first occurrence it has the incorrect value; for the other ones it has a default value
- `About to reset the video context with the user configuration`
- `The video context reset with the user configuration failed: -3`
- `About to reset the video context with the default configuration`
- `Saving the fallback/default video configuration to basic.ini`
5. `basic.ini` has the default **Video** config
6. Check that non-Video parameters has not changed.

Tested on Windows and macOS.

### Types of changes
A theoretical bug fix.

### Checklist:

- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] Some comments added to the code.

### CODE REVIEW

File: obs-studio-server/source/nodeobs_service.cpp [hunk 1]
```diff
@@ -442,9 +442,44 @@ int OBS_service::resetVideoContext(bool reload)
...
+	}
+
+	return errorcode;
+}
+
+int OBS_service::doResetVideoContext(const obs_video_info& ovi)
+{
 	try {
-		return obs_reset_video(&ovi);
+		return obs_reset_video(const_cast<obs_video_info*>(&ovi));
```
Comment (EddyGharbi) [inline obs-studio-server/source/nodeobs_service.cpp]: Why do we need to cast here?

File: obs-studio-server/source/nodeobs_service.h [hunk 2]
```diff
@@ -293,12 +293,17 @@ class OBS_service
 	// Reset contexts
 	static bool resetAudioContext(bool reload = false);
 	static int  resetVideoContext(bool reload = false);
+	static int	doResetVideoContext(const obs_video_info& ovi);
```
Comment (EddyGharbi) [inline obs-studio-server/source/nodeobs_service.h]: Extra spaces

File: obs-studio-server/source/nodeobs_service.cpp [hunk 3]
```diff
@@ -442,9 +442,44 @@ int OBS_service::resetVideoContext(bool reload)
...
+	//   D3D11CreateDevice,
+	//   etc.	
+	// OBS_VIDEO_INVALID_PARAM: A parameter is invalid.
+	// OBS_VIDEO_CURRENTLY_ACTIVE: Video is currently active.
+	// OBS_VIDEO_MODULE_NOT_FOUND: Could not load a dynamic library (ovi.graphics_module):
+	//   libobs-d3d11.dll,
+	//   libobs-opengl.
+	// OBS_VIDEO_FAIL: Generic failure.
+	if (errorcode == OBS_VIDEO_FAIL || errorcode == OBS_VIDEO_INVALID_PARAM) {
+
```
Comment (EddyGharbi) [inline obs-studio-server/source/nodeobs_service.cpp]: Extra line

File: obs-studio-server/source/nodeobs_service.cpp [hunk 4]
```diff
@@ -442,9 +442,44 @@ int OBS_service::resetVideoContext(bool reload)
...
+	//   CreateDXGIFactory1,
+	//   DXGIFactory1::EnumAdapters1,
+	//   D3D11CreateDevice,
+	//   etc.	
+	// OBS_VIDEO_INVALID_PARAM: A parameter is invalid.
+	// OBS_VIDEO_CURRENTLY_ACTIVE: Video is currently active.
+	// OBS_VIDEO_MODULE_NOT_FOUND: Could not load a dynamic library (ovi.graphics_module):
+	//   libobs-d3d11.dll,
+	//   libobs-opengl.
+	// OBS_VIDEO_FAIL: Generic failure.
```
Comment (EddyGharbi) [inline obs-studio-server/source/nodeobs_service.cpp:460]: Thanks for this, good call to add it as comments, very clear

File: obs-studio-server/source/nodeobs_service.cpp (hunk 1 above)
Comment (eugen15) [reply obs-studio-server/source/nodeobs_service.cpp]: Fixed. Explained in a code comment.

**Review by EddyGharbi — APPROVED** [review]

**Review by summeroff — APPROVED** [review]

**Pull request merged**